hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/processors/StreamToStreamJoinP.java

```diff
@@ -0,0 +1,234 @@
+     * </pre>
+     */
+    private final JetJoinInfo joinInfo;
+    private final Map<Byte, ToLongFunctionEx<JetSqlRow>> leftTimeExtractors;
+    private final Map<Byte, ToLongFunctionEx<JetSqlRow>> rightTimeExtractors;
+    private final Map<Byte, Map<Byte, Long>> postponeTimeMap;
+    private final Tuple2<Integer, Integer> columnCount;
+
+    private final Map<Byte, Map<Byte, Long>> wmState = new HashMap<>();
+
+    private ExpressionEvalContext evalContext;
+    private Iterator<JetSqlRow> pos;
+
+    private JetSqlRow currItem;
+    // NOTE: we are using LinkedList, because we are expecting:
+    // (1) removals in the middle,
+    // (2) traversing whole list without indexing.
+    private final List<JetSqlRow>[] buffer = new List[]{new LinkedList<>(), new LinkedList<>()};
+    private final Queue<JetSqlRow> pendingOutput = new ArrayDeque<>();
+    private Watermark pendingWatermark;
+
+    public StreamToStreamJoinP(
+            final JetJoinInfo joinInfo,
+            final Map<Byte, ToLongFunctionEx<JetSqlRow>> leftTimeExtractors,
+            final Map<Byte, ToLongFunctionEx<JetSqlRow>> rightTimeExtractors,
+            final Map<Byte, Map<Byte, Long>> postponeTimeMap,
+            final Tuple2<Integer, Integer> columnCount
+    ) {
+        this.joinInfo = joinInfo;
+        this.leftTimeExtractors = leftTimeExtractors;
+        this.rightTimeExtractors = rightTimeExtractors;
+        this.postponeTimeMap = postponeTimeMap;
+        this.columnCount = columnCount;
+
+        for (byte key : postponeTimeMap.keySet()) {
+            Map<Byte, Long> valueMap = postponeTimeMap.get(key);
+            Map<Byte, Long> map = new HashMap<>();
+            for (byte j : valueMap.keySet()) {
+                map.put(j, Long.MIN_VALUE);
+            }
+            wmState.put(key, map);
+        }
+    }
+
+    @Override
+    protected void init(@Nonnull Context context) throws Exception {
+        this.evalContext = ExpressionEvalContext.from(context);
+    }
+
+    @SuppressWarnings("ConstantConditions")
+    @Override
+    public boolean tryProcess(int ordinal, @Nonnull Object item) {
+        while (!pendingOutput.isEmpty()) {
+            if (!tryEmit(pendingOutput.peek())) {
+                return false;
+            } else {
+                pendingOutput.remove();
+            }
+        }
+
+        //  having side input, traverse opposite buffer
+        if (currItem == null) {
+            currItem = (JetSqlRow) item;
+            buffer[ordinal].add(currItem);
+            pos = buffer[1 - ordinal].iterator();
+
+            JetSqlRow joinedRow = null;
+            // If current join type is LEFT/RIGHT and opposite buffer is empty,
+            // we should to produce input row with null-filled opposite side.
+            if (!pos.hasNext() && !joinInfo.isInner()) {
+                if (ordinal == 1 && joinInfo.isLeftOuter()) {
+                    // fill LEFT side with nulls
+                    joinedRow = ExpressionUtil.join(
```

Comment:
Outer join should not be handled here. Even if there's no matching row in the buffers, we might receive one in the future. We need to handle outer join when a WM is received - if a row we're removing from the buffer never had a matching row, then we'll join with the null row and emit.